src/context/index.ts

```diff
@@ -0,0 +1,232 @@
+import { mapDistinct, _Region, parseRegion } from '../shared'
+import '../shared/extension'
+import { Baser } from './Baser'
+import { loadLogs } from './loadLogs'
+import { Panel } from './Panel'
+
+declare module 'vscode' {
+	interface Diagnostic {
+		_result: Result
+	}
+}
+
+export const regionToSelection = (doc: TextDocument, region: _Region | undefined) => {
+	if (!region) return new Selection(0, 0, 0, 0) // TODO: Decide if empty regions should be pre-filtered.
+	return Array.isArray(region)
+		? (region.length === 4
+			? new Selection(...region)
+			: (() => {
+				const [byteOffset, byteLength] = region
+				const startColRaw = byteOffset % 16
+				const endColRaw = (byteOffset + byteLength) % 16
+				return new Selection(
+					Math.floor(byteOffset / 16),
+					10 + startColRaw + Math.floor(startColRaw / 2),
+					Math.floor((byteOffset + byteLength) / 16),
+					10 + endColRaw + Math.floor(endColRaw / 2),
+				)
+			})())
+		: (() => {
+			const line = doc.lineAt(region)
+			return new Selection(
+				line.range.start.line,
+				line.firstNonWhitespaceCharacterIndex,
+				line.range.end.line,
+				line.range.end.character,
+			)
+		})()
+	}
+
+export class Store {
+	static extensionPath: string | undefined
+	static globalState: Memento
+
+	@observable.shallow logs = [] as Log[]
+	@computed public get results() {
+		const runs = this.logs.map(log => log.runs).flat()
+		return runs.map(run => run.results).filter(run => run).flat() as Result[]
+	}
+	@computed public get distinctArtifactNames() {
+		const fileAndUris = this.logs.map(log => [...log._distinct.entries()]).flat()
+		return mapDistinct(fileAndUris)
+	}
+
+	constructor() {
+		intercept(this.logs, objChange => {
+			const change = objChange as unknown as IArrayWillSplice<Log>
+			change.added = change.added.filter(log => this.logs.every(existing => existing._uri !== log._uri))
+			return objChange
+		})
+	}
+}
+
+export async function activate(context: ExtensionContext) {
+	const disposables = context.subscriptions
+	Store.extensionPath = context.extensionPath
+	Store.globalState = context.globalState
+	disposables.push(commands.registerCommand('sarif.clearState', () => {
+		context.globalState.update('view', undefined)
+		commands.executeCommand('workbench.action.reloadWindow')
+	}))
+	const store = new Store()
+
+	// Boot
+	const uris = await workspace.findFiles('.sarif/**/*.sarif')
+	store.logs.push(...await loadLogs(uris))
+
+	// Basing
+	const urisNonSarif = await workspace.findFiles('**/*', '.sarif') // Ignore folders?
+	const fileAndUris = urisNonSarif.map(uri => [uri.path.split('/').pop(), uri.path])  as [string, string][]
+	const baser = new Baser(mapDistinct(fileAndUris), store)
+
+	// Panel
+	const panel = new Panel(context, baser, store)
+	if (uris.length) await panel.show()
+	disposables.push(commands.registerCommand('sarif.showPanel', () => panel.show()))
```

Comment:
>disposables.push( [](start = 1, length = 17)

You can use context.subscriptions for this instead of your own disposables array.

Reply:
This is really syntactic-sugar/explanatory-variable for my own brain. "Disposables" seems to what most people call this. Not sure why VS Code thinks of these as "subscriptions" which means something different in my head.

Reply:
Line 16: const disposables = context.subscriptions

This is really syntactic-sugar/an-explanatory-variable for my own brain. "Disposables" seems to what most people call this pattern. Not sure why VS Code calls these as "subscriptions" which means something different in my head.

---
In reply to: [434584876](https://github.com/microsoft/sarif-vscode-extension/pull/279#discussion_r434584876) [](ancestors = 434584876)

Reply:
Tracking the general issue of disposables with the other comment.

---
In reply to: [436290264](https://github.com/microsoft/sarif-vscode-extension/pull/279#discussion_r436290264) [](ancestors = 436290264,434584876)